runtime-light/state/component-state.cpp

```diff
@@ -45,11 +70,37 @@ void ComponentState::parse_ini_arg(std::string_view key_view, std::string_view v
 }
 
 void ComponentState::parse_runtime_config_arg(std::string_view value_view) noexcept {
-  if (auto opt_config{json_decode(value_view)}; opt_config) [[likely]] {
-    runtime_config = *std::move(opt_config);
-  } else {
-    kphp::log::warning("runtime config isn't a valid JSON");
+  auto expected_canonicalized_path{k2::canonicalize(value_view)};
+  if (!expected_canonicalized_path) [[unlikely]] {
+    return kphp::log::warning("error canonicalizing runtime-config path -> '{}'", value_view);
+  }
+
+  const auto [runtime_config_path, runtime_config_path_size]{*std::move(expected_canonicalized_path)};
+  auto expected_runtime_config_file{kphp::fs::file::open({runtime_config_path.get(), runtime_config_path_size}, "r")};
+  if (!expected_runtime_config_file) [[unlikely]] {
+    return kphp::log::warning("error opening runtime-config: error code -> {}", expected_runtime_config_file.error());
+  }
+
+  struct stat stat {};
+  if (auto error_code{k2::stat({runtime_config_path.get(), runtime_config_path_size}, std::addressof(stat))}; error_code != k2::errno_ok) [[unlikely]] {
+    return kphp::log::warning("error getting runtime-config stat: error code -> {}", error_code);
+  }
+
+  const auto runtime_config_mem{std::unique_ptr<char, decltype(std::addressof(kphp::memory::script::free))>{
+      reinterpret_cast<char*>(kphp::memory::script::alloc(static_cast<size_t>(stat.st_size))), kphp::memory::script::free}};
+  kphp::log::assertion(runtime_config_mem != nullptr);
+  auto runtime_config_buf{std::span<char>{runtime_config_mem.get(), static_cast<size_t>(stat.st_size)}};
+  if (auto expected_read{(*expected_runtime_config_file).read(std::as_writable_bytes(runtime_config_buf))}; !expected_read || *expected_read != stat.st_size)
+      [[unlikely]] {
+    return kphp::log::warning("error reading runtime-config: error code -> {}, read bytes -> {}", expected_read.error_or(k2::errno_ok),
+                              expected_read.value_or(0));
+  }
+
+  auto opt_config{json_decode({runtime_config_buf.data(), runtime_config_buf.size()})};
+  if (!opt_config) [[unlikely]] {
```

Comment:
`has_value()`?

Reply:
It's a frequent pattern in `runtime-light`, so I'd keep it as it is